File: mobile/src/main/java/com/example/android/uamp/model/MusicProvider.java
package com.example.android.uamp.model;


import android.content.res.Resources;
import android.graphics.Bitmap;
import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.MediaMetadataCompat;

import java.util.List;

public interface MusicProvider {
    Iterable<String> getGenres();

    Iterable<MediaMetadataCompat> getShuffledMusic();

    Iterable<MediaMetadataCompat> getMusicsByGenre(String genre);

    Iterable<MediaMetadataCompat> searchMusicBySongTitle(String query);

    Iterable<MediaMetadataCompat> searchMusicByAlbum(String query);

    Iterable<MediaMetadataCompat> searchMusicByArtist(String query);

    MediaMetadataCompat getMusic(String musicId);

    void updateMusicArt(String musicId, Bitmap albumArt, Bitmap icon);

    void setFavorite(String musicId, boolean favorite);

    boolean isInitialized();

    boolean isFavorite(String musicId);

    void retrieveMediaAsync(MusicProviderImpl.Callback callback);

    List<MediaBrowserCompat.MediaItem> getChildren(String mediaId, Resources resources);
}

File: mobile/src/main/java/player/model/MusicProvider.kt
package player.model


import android.content.res.Resources
import android.graphics.Bitmap
import android.support.v4.media.MediaBrowserCompat
import android.support.v4.media.MediaMetadataCompat

interface MusicProvider {
    val genres: Iterable<String>?

    val shuffledMusic: Iterable<MediaMetadataCompat>?

    val isInitialized: Boolean

    fun getMusicsByGenre(genre: String): Iterable<MediaMetadataCompat>?

    fun searchMusicBySongTitle(query: String): Iterable<MediaMetadataCompat>?

    fun searchMusicByAlbum(query: String): Iterable<MediaMetadataCompat>?

    fun searchMusicByArtist(query: String): Iterable<MediaMetadataCompat>?

    fun getMusic(musicId: String?): MediaMetadataCompat?

    fun updateMusicArt(musicId: String?, albumArt: Bitmap, icon: Bitmap)

    fun setFavorite(musicId: String?, favorite: Boolean)

    fun isFavorite(musicId: String?): Boolean

    fun retrieveMediaAsync(callback: MusicProviderImpl.Callback)

    fun getChildren(mediaId: String, resources: Resources): List<MediaBrowserCompat.MediaItem>?
}
